test/ejabberd_sm_SUITE.erl

```diff
@@ -375,15 +395,34 @@ unload_meck() ->
     meck:unload(acl),
     meck:unload(gen_hook),
     meck:unload(ejabberd_commands),
-    meck:unload(mongoose_domain_api).
+    meck:unload(mongoose_domain_api),
+    catch ets:delete(test_c2s_info),
+    catch meck:unload(mongoose_c2s).
 
-set_test_case_meck(MaxUserSessions) ->
+set_test_case_meck(MaxUserSessions, MeckC2s) ->
+    ets:new(test_c2s_info, [public, named_table]),
     meck:new(acl, []),
     meck:expect(acl, match_rule, fun(_, _, _, _) -> MaxUserSessions end),
     meck:new(gen_hook, []),
     meck:expect(gen_hook, run_fold, fun(_, _, Acc, _) -> {ok, Acc} end),
     meck:new(mongoose_domain_api, []),
-    meck:expect(mongoose_domain_api, get_domain_host_type, fun(H) -> {ok, H} end).
+    meck:expect(mongoose_domain_api, get_domain_host_type, fun(H) -> {ok, H} end),
+    do_meck_c2s(MeckC2s).
+
+do_meck_c2s(false) ->
+    ok;
+do_meck_c2s(true) ->
+    %% Very simple mock, not even reproducing async behaviour
+    %% It is for a limited use only, for more complex tests use a real c2s process (i.e. probably big tests)
+    meck:new(mongoose_c2s, [passthrough]),
+    meck:expect(mongoose_c2s, async_with_state, fun(Pid, Fun, Args) -> apply(Fun, [{ministate, Pid}|Args]), ok end),
+    GF = fun({ministate, Pid}) -> case ets:lookup(test_c2s_info, Pid) of [] -> #{}; [{Pid, Info}] -> Info end end,
```

Comment:
Please don't inline nested expressions like this. A named helper function might be better here.